pkg/logql/log/json.go

```diff
@@ -0,0 +1,165 @@
+package log
+
+import (
+	"bytes"
+	"fmt"
+	"unicode/utf8"
+
+	"github.com/buger/jsonparser"
+	jsoniter "github.com/json-iterator/go"
+)
+
+const (
+	jsonSpacer      = '_'
+	duplicateSuffix = "_extracted"
+	trueString      = "true"
+	falseString     = "false"
+)
+
+var (
+	_                       Stage = &JSONParser{}
+	errUnexpectedJSONObject       = fmt.Errorf("expecting json object(%d), but it is not", jsoniter.ObjectValue)
+)
+
+type JSONParser struct {
+	buf    []byte // buffer used to build json keys
+	prefix []byte
+	lbs    *LabelsBuilder
+
+	keys internedStringSet
+}
+
+// NewJSONParser creates a log stage that can parse a json log line and add properties as labels.
+func NewJSONParser() *JSONParser {
+	return &JSONParser{
+		buf:    make([]byte, 0, 1024),
+		prefix: make([]byte, 0, 1024),
+		keys:   internedStringSet{},
+	}
+}
+
+func (j *JSONParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	if lbs.ParserLabelHints().NoLabels() {
+		return line, true
+	}
+
+	j.lbs = lbs
+	j.prefix = j.prefix[:0]
+	j.buf = j.buf[:0]
+	if err := jsonparser.ObjectEach(line, j.objectCallback); err != nil {
+		lbs.SetErr(errJSON)
+		return line, true
+	}
+	return line, true
+}
+
+func (j *JSONParser) objectCallback(key, value []byte, dataType jsonparser.ValueType, offset int) error {
+	switch dataType {
+	case jsonparser.String, jsonparser.Number, jsonparser.Boolean, jsonparser.Null:
+		j.handleValue(key, value, dataType)
+		return nil
+	case jsonparser.Object:
+		var ok bool
+		old := j.prefix
+		j.prefix, ok = j.appendPrefix(j.prefix, key)
+		if !ok {
+			j.prefix = old
+			return nil
+		}
+		err := jsonparser.ObjectEach(value, j.objectCallback)
+		j.prefix = old
+		return err
+	}
+	return nil
+}
+
+func (j *JSONParser) handleValue(field []byte, value []byte, dataType jsonparser.ValueType) {
+	// the first time we use the field as label key.
+	if len(j.prefix) == 0 {
+		key, ok := j.keys.Get(field, func() (string, bool) {
+			j.buf = sanitizeLabelTo(field, j.buf, true)
+			if j.lbs.BaseHas(unsafeGetString(j.buf)) {
+				j.buf = append(j.buf, duplicateSuffix...)
+			}
+			if !j.lbs.ParserLabelHints().ShouldExtract(unsafeGetString(j.buf)) {
+				return "", false
+			}
+
+			return string(j.buf), true
```

Comment:
If I read `sanitizeLabelTo` correctly it makes a copy. Won't `string(j.buf)` make another copy here?

Reply:
I think `sanitizeLabelTo` does avoid a copy, that's why it's named like this it sanitize into a buffer.